### PR TITLE
Feat: Add async streaming support to `query_engine`

### PR DESCRIPTION
**Pull request by Kigstn**
# Description

Previously, the query engine had no support for async streaming while other parts of the lib, such as the chat engine, had. 
There was already support for non-async streaming in the query engine as well.

This PR aims to remedy this missing functionality by adding support for async streaming.
None of the existing API was changed.

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [x] No, this is a `llama-index-core` change

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update 
       -> Some added guides for async streaming would be nice, but not required. Personally, I find this to be very lacking atm anyways - so updates would be appreciated!

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

#### Test Code
```py
import os
import asyncio

from llama_index.core import Settings, SimpleDirectoryReader, VectorStoreIndex

from dotenv import load_dotenv
from llama_index.embeddings.azure_openai import AzureOpenAIEmbedding
from llama_index.llms.azure_openai import AzureOpenAI

load_dotenv()


def setup():
    openai_api_key = os.environ['OPENAI_API_KEY']
    openai_api_base = os.environ['OPENAI_API_BASE']
    openai_api_version = os.environ['OPENAI_API_VERSION']
    openai_api_llm_deployment_name = os.environ['OPENAI_API_LLM_DEPLOYMENT_NAME']

    llm = AzureOpenAI(
        model='gpt-35-turbo-16k',
        deployment_name=openai_api_llm_deployment_name,
        api_key=openai_api_key,
        azure_endpoint=openai_api_base,
        api_version=openai_api_version,
    )

    openai_api_embed_model_deployment_name = os.environ['OPENAI_API_EMBED_MODEL_DEPLOYMENT_NAME']

    embedding_model = AzureOpenAIEmbedding(
        model='text-embedding-ada-002',
        deployment_name=openai_api_embed_model_deployment_name,
        api_key=openai_api_key,
        azure_endpoint=openai_api_base,
        api_version=openai_api_version,
        embed_batch_size=1
    )

    Settings.llm = llm
    Settings.embed_model = embedding_model


async def main():
    documents = SimpleDirectoryReader('test_data/paul_graham_essay').load_data()
    index = VectorStoreIndex.from_documents(documents)

    query_engine = index.as_query_engine(streaming=True)
    streaming_response = await query_engine.aquery("What did the author do growing up?")

    async for token in streaming_response.async_response_gen:
        print(token, end="")


setup()
asyncio.run(main())
```

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-core/llama_index/core/base/response/schema.py [hunk 1]
```diff
@@ -139,4 +139,66 @@ def get_formatted_sources(self, length: int = 100, trim_text: int = True) -> str
...
+        async with self._lock:
+            if self.response_txt is None and self.async_response_gen is not None:
+                self.response_txt = ""
+                async for text in self.async_response_gen:
+                    self.response_txt += text
+                    yield text
+            else:
+                yield self.response_txt
+
+    async def yield_response(self) -> TokenAsyncGen:
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/base/response/schema.py:170]: should this be renamed to `async_response_gen` to match how it works for agents/chat engines?

File: llama-index-core/llama_index/core/response_synthesizers/refine.py
Comment (logan-markewich) [inline llama-index-core/llama_index/core/response_synthesizers/refine.py:1]: I think we also need to update tree summarize

Comment (Kigstn) [reply llama-index-core/llama_index/core/response_synthesizers/refine.py:1]: [It looks like tree summarise already support streaming.](https://github.com/run-llama/llama_index/blob/main/llama-index-core/llama_index/core/response_synthesizers/tree_summarize.py#L88) But it uses non async streams even for the async calls which should really be changed. 

I'll happily implement it in this PR - but that would be a breaking change and should IMO be done in a different PR.

Comment (logan-markewich) [reply llama-index-core/llama_index/core/response_synthesizers/refine.py:1]: ah I see what you mean now. A tad sad that snuck in there

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**